src/tokenizer/mod.rs

```
@@ -1397,12 +1402,118 @@ impl<Sink: TokenSink> Tokenizer<Sink> {
+
+    pub use super::interface::{Token, DoctypeToken, TagToken, CommentToken};
+    pub use super::interface::{CharacterTokens, NullCharacterToken, EOFToken, ParseError};
+    pub use super::interface::{Doctype, Attribute, TagKind, StartTag, EndTag, Tag};
+
+    use super::buffer_queue::{BufferQueue};
+    use std::mem::replace;
+    use std::borrow::Cow::{self, Borrowed};
+
+    // TokenMatch implements the TokenSink trait. It is used for testing to see 
+    // if current_line is being updated when process_token is called. The lines
+    // vector is a collection of the line numbers that each token is on.
+    struct TokenMatch {
+        tokens: Vec<Token>,
+        current_str: StrTendril,
+        exact_errors: bool,
+        lines: Vec<u64>,
+    }
+    
+    impl TokenMatch {
+        fn new(exact_errors: bool) -> TokenMatch {
+            TokenMatch {
+                tokens: vec!(),
+                current_str: StrTendril::new(),
+                exact_errors: exact_errors,
+                lines: vec!(),
+            }
+        }
+
+        fn push(&mut self, token: Token) {
+            self.finish_str();
+            self.tokens.push(token);
+        }
+
+        fn finish_str(&mut self) {
+            if self.current_str.len() > 0 {
+                let s = replace(&mut self.current_str, StrTendril::new());
+                self.tokens.push(CharacterTokens(s));
+            }
+        }
+
+    }
+
+    impl TokenSink for TokenMatch {
+        
+        type Handle = ();
+
+        fn process_token(&mut self, token: Token, line_number: u64) -> TokenSinkResult<Self::Handle>{
+            
+            self.lines.push(line_number);
+
+            match token {
+                CharacterTokens(b) => {
+                    self.current_str.push_slice(&b);
+                }
+
+                NullCharacterToken => {
+                    self.current_str.push_char('\0');
+                }
+
+                ParseError(_) => if self.exact_errors {
```

We can use `ParseError(_) => panic!("unexpected parse error")` for this test.